crates/iota-indexer/src/read.rs

```diff
@@ -584,28 +645,123 @@ impl IndexerReader {
         })
     }
 
-    pub fn get_checkpoints(
+    pub async fn get_checkpoints(
         &self,
         cursor: Option<u64>,
         limit: usize,
         descending_order: bool,
     ) -> Result<Vec<iota_json_rpc_types::Checkpoint>, IndexerError> {
-        self.get_checkpoints_from_db(cursor, limit, descending_order)?
+        let checkpoints = self
+            .spawn_blocking(move |this| {
+                this.get_checkpoints_from_db(cursor, limit, descending_order)?
+                    .into_iter()
+                    .map(iota_json_rpc_types::Checkpoint::try_from)
+                    .collect::<Result<Vec<_>, IndexerError>>()
+            })
+            .await?;
+
+        // resolve the expected range of checkpoint sequence numbers
+        let expected_range: Vec<CheckpointSequenceNumber> = match (cursor, descending_order) {
+            // ascending from 0: expect [0, 1, ..., limit-1].
+            (None, false) => {
+                // if genesis is present, data is not pruned, use actual results length which
```

Comment:
as below, if genesis is present then that's it, the data is complete, we will not have situation where genesis is there but sth else is pruned

the expected range for fallback is always "(0..limit as u64).collect()", no matter how many checkpoints are returned from DB